### PR TITLE
fix: ryu patching is cleaned up

### LINKED ISSUE #13687: ryu patching is understood and fixed for 1.8
### Problem
ryu is not patched with magma specific QFI changes in master and the 1.8 release

#### What patches are there?
* `lte/gateway/deploy/roles/magma/files/patches` contains four patches
* the patch `lte/gateway/deploy/roles/magma/files/patches/ryu_ipfix_args.patch` seems to be applied to the checked-in files `lte/gateway/deploy/roles/magma/files/nx_actions.py` and `lte/gateway/deploy/roles/magma/files/nx_actions_3.5.py` for python `3.8` and `3.5`.
  * note: there is also `cwf/gateway/deploy/roles/ovs/files/nx_actions.py` which seems to differ from the file above
* in the same way the patch `lte/gateway/deploy/roles/magma/files/patches/0001-Set-unknown-dpid-ofctl-log-to-debug.patch` is pre-applied to `lte/gateway/deploy/roles/magma/files/service.py`.
* note: multiple code formattings were applied to the pre-patched files - this makes it hard to see what patch was actually applied
  * verified manually that `lte/gateway/deploy/roles/magma/files/nx_actions.py` without any formatting changes coincides to a freshly patched `nx_actions.py` from the official 4.34 repo.
* the other patches (`0002-`-`0003-`) are not pre-applied to any files.

##### for the vm - provisioning
* for the magma dev VM ryu is patch during provisioning - see `lte/gateway/deploy/roles/magma/tasks/main.yml:424`+
* ryu seems not to be installed during provisioning for a long time, i.e., patching does not really make sense here
* additionally `ryu_ipfix_args.patch` is applied by overriding `nx_actions.py` by the pre-applied patched files (see above) where `3.5` and `3.8` seems to be mixed up
* the other patches are not applied - the patch files would overriding ryu files (if ryu was installed) and by this damaging ryu

##### for the vm - build
* see `orc8r/gateway/python/python.mk:56`+
* triggered via `make run`
* this is why the patches are recognized in the integration tests
* note: `0001-Set-unknown-dpid-ofctl-log-to-debug.patch` is applied to the wrong file and is failing

##### via bazel
* see `bazel/python_repositories.bzl`
* this is not yet used in production builds

##### for magma.deb
* `nx_actions_3.5.py` ans `service.py` (see above) are deployed (but not `nx_actions.py`) and they override the respective files in ryu by `lte/gateway/release/magma-postinst`.
* This is, ryu is installed without patches and the post-install script applies (two of the four) patches by overriding ryu files by pre-patched files.

##### for containerized AGW
* done in `lte/gateway/docker/services/python/Dockerfile:71` via `make buildenv`
* relevant patches are applied via `orc8r/gateway/python/python.mk:56` (not on `service.py` but this might be not that relevant)
* checked with @jheidbrink and @ajahl 

#### Recommendations for moving forward
* for 1.8
  * manually build a patched ryu package and add it to the magma artifactory
    * make sure this is not replaced by a native ubuntu package
* not for 1.8
  * remove the ansible patching - this is just damaged
  * remove the pre-patched files and use a patched ryu .deb package or apply the patches during a build

### PR DESCRIPTION
**Pull request by nstng**
## Summary

This change originates from the findings in #13687. Best reviewed commit-by-commit.

Goal:
* production .deb environments install a patched `python3-ryu.deb` and do not use pre-patched ryu files.
* cwf docker builds use a patched `python3-ryu.deb` and do not use pre-patched ryu files.
* magma_dev VM environments patch ryu by applying the patches via `python.mk` `py_patches:`
  * no change needed (only fix to apply patch to correct file)
* agw docker builds patch ryu by applying the patches via `python.mk` `py_patches:`
  * no change needed here

- [x] focal-ci needs to provide a patched `python3-ryu.deb` created by the script `ryu_build.sh` in this PR - initiated by `~/magma/third_party/build$ ./build.py ryu`.

## Test Plan

- [x] cwf docker build and tests work
  - https://github.com/nstng/magma/actions/runs/2946313775
- [x] `~/magma/third_party/build$ ./build.py ryu` builds correct ryu package
- [ ] integ tests - not really a change here (only additional patch now applied correctly - only changes log level of one statement)

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Comment by nstng** [comment]
@koolzz @pshelar please let me know if there was a good reason for using pre-patched files to apply ryu patches in certain situations, i.e., if the change here should be dropped.

**Review by crasu — APPROVED** [review]

File: third_party/build/bin/ryu_build.sh [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# package ryu with magma patches
+
+set -e
```
Comment (LKreutzer) [inline third_party/build/bin/ryu_build.sh:16]: - Might be good to add `set -euo pipefail`.
- Is this script just for documentation or will this be the standard way to create the package?

File: third_party/build/bin/ryu_build.sh [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+SCRIPT_DIR="$(dirname "$(realpath "$0")")"
+# shellcheck source=/dev/null # relative path unknown and option -x not given
+source "$SCRIPT_DIR"/../lib/util.sh
+PKGVERSION=4.34
+ITERATION=1
+# shellcheck disable=SC2034 # variable is used in third_party/build/lib/util.sh
+VERSION="$PKGVERSION-$ITERATION"
+PKGNAME=python3-ryu
```
Comment (LKreutzer) [inline third_party/build/bin/ryu_build.sh:24]: Is it too late or is there another reason not to call this package `python3-ryu-magma`?

File: third_party/build/bin/ryu_build.sh [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  OUTPUT_DIR=$(pwd)
+else
+  OUTPUT_DIR=$1
+  if [ ! -d "$OUTPUT_DIR" ]; then
+    echo "error: $OUTPUT_DIR is not a valid directory. Exiting..."
+    exit 1
+  fi
+fi
+
+# remove privious working directory
```
Comment (LKreutzer) [inline third_party/build/bin/ryu_build.sh]: Typo: "previous"

**Review by LKreutzer — CHANGES_REQUESTED** [review]

File: third_party/build/bin/ryu_build.sh (hunk 1 above)
Comment (nstng) [reply third_party/build/bin/ryu_build.sh:16]: have you tried if this works in the current setup, i.e., `/third_party/build$ ./build.py ryu`?

File: third_party/build/bin/ryu_build.sh (hunk 2 above)
Comment (nstng) [reply third_party/build/bin/ryu_build.sh:24]: the name is created via `build.py` and `utils.sh`.

File: third_party/build/bin/ryu_build.sh (hunk 1 above)
Comment (LKreutzer) [reply third_party/build/bin/ryu_build.sh:16]: No, I don't have a prod VM (on the magma_vm the `ryu_build.sh` does not exist).

Comment (nstng) [reply third_party/build/bin/ryu_build.sh:16]: discussed offline - does not work with util.sh. Refactoring is out of scope here.

File: third_party/build/bin/ryu_build.sh (hunk 2 above)
Comment (nstng) [reply third_party/build/bin/ryu_build.sh:24]: as above

File: third_party/build/bin/ryu_build.sh (hunk 3 above)
Comment (nstng) [reply third_party/build/bin/ryu_build.sh]: done

**Review by LKreutzer — APPROVED** [review]

**Comment by sebathomas** [comment]
The changes look good, just wondering about the cwf test in the test plan. Did you manage to get it green in the CI?

**Review by ajahl — APPROVED** [review]

**Comment by nstng** [comment]
> The changes look good, just wondering about the cwf test in the test plan. Did you manage to get it green in the CI?

@sebathomas Here is a successful run - not sure why the "Unit Test Results" Job is displayed as failed (or where this job actually comes from), iiuc then the runs were flaky but at least one test run was successful.
https://github.com/nstng/magma/actions/runs/2994969304

**Comment by nstng** [comment]
> @koolzz @pshelar please let me know if there was a good reason for using pre-patched files to apply ryu patches in certain situations, i.e., if the change here should be dropped.

I will merge this for now - let me know if you have post-feed-back for this issue.

**Pull request merged**